gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+		_, prefixDst, prefixBackupBase := f.entryNames(dstRoot, "", prefix, backupDir)
+
+		prefixBackupName := prefixBackupBase + ".backup"
+		if osutil.FileExists(prefixBackupName) {
+			continue
+		}
+		if !osutil.IsDirectory(prefixDst) {
+			// does not exist now, will be created on the fly and
+			// removed during rollback
+			continue
+		}
+		if err := os.MkdirAll(filepath.Dir(prefixBackupName), 0755); err != nil {
+			return fmt.Errorf("cannot create backup prefix: %v", err)
+		}
+		if err := makeStamp(prefixBackupName); err != nil {
+			return fmt.Errorf("cannot create a checkpoint for directory: %v", err)
+		}
+	}
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+
+	srcName, dstName, backupBase := f.entryNames(dstRoot, source, target, backupDir)
+
+	backupName := backupBase + ".backup"
+	sameStamp := backupBase + ".same"
+	preserveStamp := backupBase + ".preserve"
+
+	if osutil.FileExists(backupName) || osutil.FileExists(sameStamp) {
+		// file already checked, move on
+		return nil
+	}
+
+	if strutil.SortedListContains(preserveInDst, dstName) {
+		if !osutil.FileExists(dstName) {
+			// preserve, but does not exist, will be deployed anyway
+			return nil
+		}
+		if osutil.FileExists(preserveStamp) {
+			// already stamped
+			return nil
+		}
+		// make a stamp
+		if err := makeStamp(preserveStamp); err != nil {
+			return fmt.Errorf("cannot create preserve stamp: %v", err)
+		}
+		return nil
+	}
+
+	orig, err := os.Open(dstName)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot open destination file: %v", err)
+		}
+
+		// destination does not exist, no need for backup
+		return f.backupOrCheckpointPrefix(dstRoot, target, backupDir)
+	}
+
+	// backup of the original content
+	backup, err := makeStampFile(backupName)
+	if err != nil {
+		return fmt.Errorf("cannot create backup file: %v", err)
+	}
+	// becomes a noop if canceled
+	defer backup.Commit()
+
+	// checksum the original data while it's being copied
+	origHash := crypto.SHA1.New()
+	htr := io.TeeReader(orig, origHash)
+
+	_, err = io.Copy(backup, htr)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot backup original file: %v", err)
+	}
+
+	// digest of the update
+	updateDigest, _, err := osutil.FileDigest(srcName, crypto.SHA1)
+	if err != nil {
+		backup.Cancel()
+		return fmt.Errorf("cannot checksum update file: %v", err)
+	}
+	// digest of the currently present data
+	origDigest := origHash.Sum(nil)
+
+	if bytes.Equal(origDigest, updateDigest) {
+		// files are identical, no backup needed
+		if err := makeStamp(sameStamp); err != nil {
+			return fmt.Errorf("cannot create a checkpoint file: %v", err)
+		}
+
+		// makes the previous commit a noop
+		backup.Cancel()
+		return nil
+	}
+
+	return f.backupOrCheckpointPrefix(dstRoot, target, backupDir)
```

Comment:
again,  I probably slighty missing what this function does really, either the lack of comment or bad name for it. Is it really just checkpointing in reality?  when does it backup something?

Reply:
Yeah, it should probably be called something like `checkpointPrefix()`. It's only purpose is to create a stamp for each path component that exist in the destination, so that we can track and clean up parts of the tree that were present before the update.